firmware/src/networking/protocol/mod.rs

```diff
@@ -0,0 +1,68 @@
+//! THIS MODULE IS WIP AND NOT FINISHED YET
+//!
+//! Manages the slimevr firmware protocol. All (de)serialization and state associated with
+//! the protocol happens in here.
+#![allow(dead_code)]
+
+mod serialize;
+
+use crate::{Quat, Unreliable};
+
+use embassy_sync::{
+	blocking_mutex::raw::NoopRawMutex,
+	channel::{Channel, Receiver, Sender},
+};
+use firmware_protocol::{CbPacket, SbPacket};
+
+pub struct ImuHandle {
+	signal: &'static Unreliable<Quat>,
+}
+impl ImuHandle {
+	/// Send the latest quaternion. May drop the data.
```

Comment:
Comment implies receiver/protocol task will be in charge of throttling IMU sample rate. This is probably fine (except for code separation) when IMU report rate and slimevr sample rate do not match, but may produce odd aliasing artifacts when IMU can produce the wanted 100tps.

For example, MPUxxxx devices have configurable report rate, but closest to SlimeVR default 100tps is 92hz and 184hz. This will lead to unavoidable data loss and "incorrectly" reported IMU samples, but too bad 🤷 However, on IMUs that can provide exactly 100hz report rate, there will be slight timing issues caused by slight poll rate jitter (essentially same as tearing on non-vsynced monitors)

This will also not work for variable report rates (IMU initiates faster send rate on significant motion detection, slows down when idle)

Reply:
This function gets called from the IMU task, so the imu task gets to decide how often to call it. If the imu task decides there is no significant motion it can totally call it at a lower hz. The only thing is that we do not *guarantee* that the data will always be observed by the networking task, such as when the networking task is busy rn.

Allowing data to be dropped is still important, because we only care about the latest quat, having a buffer of even one (like what would happen if we used a channel) would not be ideal.

Note that signals still would allow us to provide an api to the imu task to observe if the current data had been observed yet: https://docs.embassy.dev/embassy-sync/git/default/signal/struct.Signal.html#method.signaled
I'm not convinced that this is a good idea though.

Reply:
Signal is still a buffer of one, just without back pressure. I disagree that the protocol implementation should decide on send rate. This either limits how adaptive and efficient we can be about sending data (forced to certain send rate by protocol), or then protocol task will also have to analyse and filter IMU data (spaghetti)

Reply:
> forced to certain send rate by protocol

the signal means an upper bound on send rate set by the imu task. If we don't send data from the imu task the control task/networking task won't get that data.

The backpressure is the exact issue though - we have no way to replace the channel with new more up to date data if we want to.

Reply:
Please note though that this code is WIP I am not tied to a particular choice of datastructure just yet though.